### PR TITLE
authentication: bind current user in the context + cleanup

### PR DESCRIPTION
**Pull request by bajtos**
- context: `add @inject.getter`

    Implement a new decorator for injecting a setter function for (re)binding a new value for the given key.

    This is useful when implementing Actions that are contributing new Elements. Compared to existing `bindElement`, this new setter approach leads to a less coupled code, because the action invoking the setter does not need to know about the key associated with the binding, in fact a setter function can be used without any Context/Binding classes. This also makes Actions easier to test in isolation.

- context: add types `Setter<T>` and `Getter<T>`

    Add two new types to make it easier to use `@inject.getter` and `@inject.setter` in TypeScript codebase. These two types represent the function that's being injected.

    Also add API docs for both the two new types, but also for `@inject.getter` and `@inject.setter`.

- authentication: bind current user in the context

    The meat of this pull request, improving the authentication component to bind the current user after it was authenticated.

     This commit also cleans the authentication component a bit more.

A follow-up for #492, connect to #483

cc @bajtos @raymondfeng @ritch @superkhau

### CODE REVIEW

File: packages/context/src/inject.ts [hunk 1]
```diff
@@ -102,17 +102,68 @@ export function inject(
...
+   *
+   * This is useful e.g. when implementing Actions that are contributing
+   * new Elements.
+   *
+   * See also `Setter<T>`.
+   *
+   * @param bindingKey The key of the value we want to set.
+   * @param metadata Optional metadata to help the injection
+   */
+  export const setter = function injectGetter(
```
Comment (raymondfeng) [inline packages/context/src/inject.ts]: `injectSetter`?

Comment (bajtos) [inline packages/context/src/inject.ts]: Good catch!

File: packages/context/src/inject.ts [hunk 2]
```diff
@@ -102,17 +102,68 @@ export function inject(
...
+   * the action is instantiated for Sequence constructor, but some
+   * of action's dependencies become bound only after other actions
+   * have been executed by the sequence.
+   *
+   * See also `Getter<T>`.
+   *
+   * @param bindingKey The key of the value we want to eventually get.
+   * @param metadata Optional metadata to help the injection
+   */
   export const getter = function injectGetter(
```
Comment (raymondfeng) [inline packages/context/src/inject.ts]: Does it represent a deferred injection?

**Comment by raymondfeng** [comment]
@bajtos I have reservations to use `@inject.setter` to allow bound classes to add bindings to the context. It's not `injection` any more as the constructor parameter or property does NOT receive any injection from the context. `@bind` is probably a better name. 

IMO, the requirement is more related to lifecycle support. If we allow bound classes to have methods to be decorated as `@start` or `@stop`, such methods can be used to bind more keys.

For `@inject.getter`, maybe `@inject(key, {deferred: true}, ..` is good enough. With `deferred: true`, `inject` does not aggressively resolve a promise. It ends up injecting the promise itself instead of the value.

**Comment by bajtos** [comment]
@raymondfeng welcome back and thank you for thoughtful comments!

> For @inject.getter, maybe @inject(key, {deferred: true}, .. is good enough. With deferred: true, inject does not aggressively resolve a promise. It ends up injecting the promise itself instead of the value.

`@inject.getter` was implemented in #492. Let me cross-post additional information from that pull request:

> implement a new @inject.getter decorator, which injects a getter function to obtain the bound value. This is useful when implementing Actions, where the action is instantiated for Sequence constructor, but some of action's dependencies become bound only after other actions have been executed by the sequence. (autenticate actions is a prime example.)

Injecting a promise itself will not work. Here is a simplified description of the execution flow in a Sequence:

 1. resolve `findRoute` (inject dependencies)
 2. resolve `authenticate` (inject dependencies like `BindingKeys.Authentication.STRATEGY`)
    - 2.1. STRATEGY provider needs METADATA
      - 2.1.1 METADATA provider needs two values that are not bound yet:
         - `BindingKeys.Context.CONTROLLER_CLASS`  
         - `BindingKeys.Context.CONTROLLER_METHOD_NAME`
 3. execute `sequence.handle`
    - 3.1 execute `findRoute()`
      - 3.1.1 `findRoute` binds CONTROLLER_CLASS and CONTROLLER_METHOD_NAME
 4. execute `authenticate()`

IIUC, in your proposal, we will inject a promise. The trouble with the promise is that the actual work has to start immediately when the promise is created, i.e. in the step 2. However, at this time, the dependencies are not defined yet. So, how and when is the promise going to be resolved? When CONTROLLER_CLASS and CONTROLLER_METHOD_NAME are finally bound? How are we going to detect that fact? What's even more important, how are we going to propagate the resolution of deferred dependencies backwards through the dependency chain (authenticate waits for strategy, it waits for metadata, it waits for controller class/method-name).

> IMO, the requirement is more related to lifecycle support. If we allow bound classes to have methods to be decorated as @start or @stop, such methods can be used to bind more keys.

I don't think `@start` and `@stop` will be enough. Each Sequence action can contribute new bindings. `findRoute` contributes controller class and method-name. `authenticate` contributes the current user. Let's consider writing an`authorize` action that will take the current user and check that the user is authorized to perform the requested action. Its dependency on the current user must be resolved/injected only after the current user has been bound by `authenticate`. In other words, we have three lifecycle points now:
 1. the sequence class instance is constructed
 2. the route is found
 3. the current user is determined

And we have code that needs to run at point 2 (the route was found, we can authenticate now) and different code to run at point 3 (the current user is known, we can authorize now). How do you envision to support this with a single `@start` phase?

**Comment by bajtos** [comment]
@raymondfeng 

> I have reservations to use `@inject.setter` to allow bound classes to add bindings to the context. It's not `injection` any more as the constructor parameter or property does NOT receive any injection from the context. `@bind` is probably a better name. 

I disagree. The constructor is receiving an injection from the context - it's a function for setting context values (bindings) that has the current context stored in its closure. I picked the name `inject.setter` because it feels consistent with `inject.getter` and at least to me, it expresses the intent very clear: inject an argument value that is a setter function.

 - `@inject('key')` says "inject the bound value"
 - `@inject.getter('key')` says "inject a function for getting the bound value"
 - `@inject.setter('key')` says "inject a function for setting the bound value"

Having said that, I am happy to discuss better names.

I find the name `@bind` misleading. If `@inject` is decorating an argument that should be injected, then I expect `@bind` to decorate an argument that should be bound. Bound to what? To a binding key? That would make sense to me for a return value or an "out" argument that's in C# but not in JavaScript/TypeScript. In LB Context, we are not binding function arguments, we are binding values to keys.

Thoughts?

File: packages/context/src/inject.ts (hunk 2 above)
Comment (bajtos) [inline packages/context/src/inject.ts]: See https://github.com/strongloop/loopback-next/pull/494#issuecomment-321175767

**Comment by raymondfeng** [comment]
@bajtos Thank you for the explanation. Here is my summary:

1. inject: consume the dependency immediately upon the instantiation 
2. inject.getter: consume the dependency later after the instantiation
3. inject.setter: produce the dependency later after the instantiation

Based on the PR, getter is async while setter is sync, is it on purpose?

**Comment by raymondfeng** [comment]
OK, I guess for setter, the value can be a Promise if needed.

**Comment by raymondfeng** [comment]
If we go down the path to fully express binding keys that an `Action` consumes (eagerly or lazily) and produces, we can use such information to calculate the dependency graph and create a sequence accordingly.

**Comment by raymondfeng** [comment]
For the PR itself, let's merge it.

**Pull request merged**